site/bps/BP-35-128-bits-support.md

```diff
@@ -0,0 +1,399 @@
+---
+title: "BP-34: 128 bits support"
+issue: https://github.com/apache/bookkeeper/603
+state: "Under Discussion"
+design doc: https://docs.google.com/document/d/1cu54dNSV2ZrdWCi40LcyX8NxXGRCW0609T_ewmK9BWM
+release: "4.9.0"
+---
+
+### Motivation
+
+BookKeeper coordinates with a metadata store to generate a cluster wide `ledgerId`.
+Currently this is a signed `64 bit` number (effectively 63 bits). This method works
+great because we have a centralized metadata store for coordinating the id generation.
+However this method may not scale as the cluster size and number of ledgers grow.
+
+[Universally unique identifier - Wikipedia](https://en.wikipedia.org/wiki/Globally_unique_identifier)
+is a preferred way to generate decentralized globally unique IDs and it takes `128 bits`.
+This method can scale well as it doesn't need a centralized coordination. 
+
+This BP proposes the changes for increasing ledger id from `63 bits` to `128 bits`.
+
+### 128 bits
+
+Since there is no native support for `128 bits` in both Java and
+[Protobuf](https://github.com/google/protobuf/issues/2180), we have to break `128 bits`
+into 2 `64 bits` numbers for representing the `128 bits` id:
+
+- ledger-id-msb: the most significant 64 bits, bit 64 - 127
+- ledger-id-lsb: the least significant 64 bits, bit 0 - 63
+
+For backward compatibility, the `ledger-id-lsb` is the current `64 bits` ledger-id.
+The `ledger-id-msb` will be added as a new field in both API and protocol. 
+
+I am proposing calling `ledger-id-msb` as `ledger-scope-id`. So the current 64bits `ledgerId` and
+the newly introduced 64bits `ledgerScopeId` together will be forming the new `128 bits` ledger id.
+
+The default `ledgerScopeId` is `0`. That means any ledgers created prior to this change are allocated
```

Comment:
> If you do go down the of generating scope from the UUID, the scope of 0 should be reserved (i.e. if UUID.randomUUID gives back a uuid with msb 0, discard and try again).

I am not sure we need to reserve scope 0 or Long.MAX_VALUE. because currently bookkeeper allows application managing id generation. 

for certain (or default) metadata driver implementation, we can reserve those scope ids if needed.

Reply:
sure.